File: packages/core/src/verification/verifyAuthorization.ts
import type { AuthorizationV1 } from "../types/authorization.js";
import type { KeySet } from "../types/keyset.js";
import type { VerificationResult, VerificationViolation } from "./types.js";
import {
  SIGNING_DOMAINS,
  findKeyInKeySets,
  keyIsActiveAt,
  signEd25519,
  verifyEd25519,
  verifyHmacDomain
} from "../crypto/signatures.js";

/** @public */
export type VerifyAuthorizationOptions = {
  now?: number;
  expectedIssuer?: string;
  expectedAudience?: string;
  expectedPolicyId?: string;
  consumedAuthIds?: readonly string[];
  trustedKeySets?: KeySet | readonly KeySet[];
  requireSignatureVerification?: boolean;
  legacyHmacSecret?: string;
};

function sortViolations(violations: VerificationViolation[]): VerificationViolation[] {
  return [...violations].sort((a, b) => {
    if (a.code < b.code) return -1;
    if (a.code > b.code) return 1;
    return (a.index ?? 0) - (b.index ?? 0);
  });
}

function hasText(value: unknown): value is string {
  return typeof value === "string" && value.length > 0;
}

function nowOrThrow(now: number | undefined): number {
  if (now !== undefined) return now;
  return Math.floor(Date.now() / 1000);
}

/** @public */
export function authorizationSigningPayload(auth: AuthorizationV1): Omit<AuthorizationV1, "signature"> {
  return {
    auth_id: auth.auth_id,
    issuer: auth.issuer,
    audience: auth.audience,
    intent_hash: auth.intent_hash,
    state_hash: auth.state_hash,
    policy_id: auth.policy_id,
    decision: auth.decision,
    issued_at: auth.issued_at,
    expiry: auth.expiry,
    alg: auth.alg,
    kid: auth.kid,
    nonce: auth.nonce,
    capability: auth.capability
  };
}

/** @public */
export function signAuthorizationEd25519(
  auth: Omit<AuthorizationV1, "signature" | "alg"> & { alg?: "Ed25519" },
  privateKeyPem: string
): AuthorizationV1 {
  const unsigned: AuthorizationV1 = {
    ...auth,
    alg: "Ed25519",
    signature: ""
  };
  const signature = signEd25519(SIGNING_DOMAINS.AUTH_V1, authorizationSigningPayload(unsigned), privateKeyPem);
  return { ...unsigned, signature };
}

/** @public */
export function verifyAuthorization(
  auth: AuthorizationV1,
  opts?: VerifyAuthorizationOptions
): VerificationResult {
  const violations: VerificationViolation[] = [];
  const now = nowOrThrow(opts?.now);
  const consumed = new Set(opts?.consumedAuthIds ?? []);

  if (!hasText(auth.decision) || auth.decision !== "ALLOW") {
    violations.push({ code: "AUTH_DECISION_INVALID", message: "authorization decision must be ALLOW" });
  }
  if (!hasText(auth.intent_hash)) {
    violations.push({ code: "AUTH_MISSING_FIELD", message: "intent_hash is required" });
  }
  if (!hasText(auth.state_hash)) {
    violations.push({ code: "AUTH_MISSING_FIELD", message: "state_hash is required" });
  }
  if (!hasText(auth.policy_id)) {
    violations.push({ code: "AUTH_MISSING_FIELD", message: "policy_id is required" });
  }
  if (!hasText(auth.issuer)) {
    violations.push({ code: "AUTH_MISSING_FIELD", message: "issuer is required" });
  }
  if (!hasText(auth.audience)) {
    violations.push({ code: "AUTH_MISSING_FIELD", message: "audience is required" });
  }
  if (!hasText(auth.auth_id)) {
    violations.push({ code: "AUTH_MISSING_FIELD", message: "auth_id is required" });
  }
  if (!Number.isInteger(auth.issued_at)) {
    violations.push({ code: "AUTH_MISSING_FIELD", message: "issued_at must be integer unix seconds" });
  }
  if (!hasText(auth.alg)) {
    violations.push({ code: "AUTH_MISSING_FIELD", message: "alg is required" });
  }
  if (!hasText(auth.kid)) {
    violations.push({ code: "AUTH_MISSING_FIELD", message: "kid is required" });
  }
  if (!hasText(auth.signature)) {
    violations.push({ code: "AUTH_MISSING_FIELD", message: "signature is required" });
  }

  if (!Number.isInteger(auth.expiry)) {
    violations.push({ code: "AUTH_MISSING_FIELD", message: "expiry must be integer unix seconds" });
  } else if (now >= auth.expiry) {
    violations.push({ code: "AUTH_EXPIRED", message: "authorization has expired" });
  }

  if (opts?.expectedIssuer !== undefined && auth.issuer !== opts.expectedIssuer) {
    violations.push({ code: "AUTH_ISSUER_MISMATCH", message: "issuer does not match expectedIssuer" });
  }
  if (opts?.expectedAudience !== undefined && auth.audience !== opts.expectedAudience) {
    violations.push({ code: "AUTH_AUDIENCE_MISMATCH", message: "audience does not match expectedAudience" });
  }
  if (opts?.expectedPolicyId !== undefined && auth.policy_id !== opts.expectedPolicyId) {
    violations.push({ code: "AUTH_POLICY_ID_MISMATCH", message: "policy_id does not match expectedPolicyId" });
  }
  if (hasText(auth.auth_id) && consumed.has(auth.auth_id)) {
    violations.push({ code: "AUTH_REPLAY", message: "auth_id has already been consumed" });
  }

  const payload = authorizationSigningPayload(auth);

  const trustedRaw = opts?.trustedKeySets;
  const trusted = trustedRaw
    ? (Array.isArray(trustedRaw) ? trustedRaw : [trustedRaw])
    : [];
  const requireSig = opts?.requireSignatureVerification ?? false;
  const hasSigMaterial = hasText(auth.alg) && hasText(auth.kid) && hasText(auth.signature);

  if (hasSigMaterial) {
    if (auth.alg === "Ed25519") {
      if (trusted.length === 0) {
        if (requireSig) {
          violations.push({ code: "AUTH_TRUST_MISSING", message: "trustedKeySets required for Ed25519 verification" });
        }
      } else {
        const key = findKeyInKeySets(trusted, auth.issuer, auth.kid, "Ed25519");
        if (!key) {
          violations.push({ code: "AUTH_KID_UNKNOWN", message: "kid not found for issuer/alg" });
        } else if (!keyIsActiveAt(key, now)) {
          violations.push({ code: "AUTH_KEY_INACTIVE", message: "key is not active at verification time" });
        } else if (!verifyEd25519(SIGNING_DOMAINS.AUTH_V1, payload, auth.signature, key.public_key)) {
          violations.push({ code: "AUTH_SIGNATURE_INVALID", message: "signature verification failed" });
        }
      }
    } else if (auth.alg === "HMAC-SHA256") {
      if (opts?.legacyHmacSecret) {
        if (!verifyHmacDomain(SIGNING_DOMAINS.AUTH_V1, payload, auth.signature, opts.legacyHmacSecret)) {
          violations.push({ code: "AUTH_SIGNATURE_INVALID", message: "legacy HMAC signature verification failed" });
        }
      } else if (requireSig) {
        violations.push({ code: "AUTH_TRUST_MISSING", message: "legacyHmacSecret required for HMAC verification" });
      }
    } else {
      violations.push({ code: "AUTH_ALG_UNSUPPORTED", message: "unsupported signature algorithm" });
    }
  }

  if (violations.length > 0) {
    return {
      ok: false,
      status: "invalid",
      violations: sortViolations(violations),
      policyId: hasText(auth.policy_id) ? auth.policy_id : undefined,
      stateHash: hasText(auth.state_hash) ? auth.state_hash : undefined
    };
  }

  return {
    ok: true,
    status: "ok",
    violations: [],
    policyId: auth.policy_id,
    stateHash: auth.state_hash
  };
}

File: examples/rust-verifier/src/verify_authorization.rs
use base64::{engine::general_purpose::STANDARD, Engine as _};
use ed25519_dalek::{Signature, Verifier};
use serde_json::Value;

use crate::canonical::authorization_signing_input;
use crate::keyset::resolve_ed25519_key;
use crate::types::{AuthorizationV1, KeySet, VerificationResult};

fn required_non_empty(v: &str) -> bool {
    !v.trim().is_empty()
}

pub fn verify_authorization(
    auth: &AuthorizationV1,
    keyset: &KeySet,
    expected_audience: &str,
    now_unix: i64,
) -> VerificationResult {
    if !required_non_empty(&auth.auth_id)
        || !required_non_empty(&auth.issuer)
        || !required_non_empty(&auth.audience)
        || !required_non_empty(&auth.intent_hash)
        || !required_non_empty(&auth.state_hash)
        || !required_non_empty(&auth.policy_id)
        || !required_non_empty(&auth.alg)
        || !required_non_empty(&auth.kid)
        || !required_non_empty(&auth.signature)
    {
        return VerificationResult::invalid("AUTH_MISSING_FIELD", "one or more required fields are missing");
    }

    if auth.decision != "ALLOW" {
        return VerificationResult::invalid("AUTH_DECISION_INVALID", "authorization decision must be ALLOW");
    }

    if auth.expiry <= now_unix {
        return VerificationResult::invalid("AUTH_EXPIRED", "authorization has expired");
    }

    if auth.audience != expected_audience {
        return VerificationResult::invalid("AUTH_AUDIENCE_MISMATCH", "audience does not match expectedAudience");
    }

    let key = match resolve_ed25519_key(keyset, &auth.issuer, &auth.kid, &auth.alg) {
        Ok(k) => k,
        Err(err) => {
            let (code, msg) = err
                .split_once(": ")
                .map(|(c, m)| (c, m))
                .unwrap_or(("AUTH_VERIFICATION_ERROR", err.as_str()));
            return VerificationResult::invalid(code, msg);
        }
    };

    let mut payload = match serde_json::to_value(auth) {
        Ok(v) => v,
        Err(_) => {
            return VerificationResult::invalid("AUTH_MALFORMED", "authorization payload serialization failed")
        }
    };

    if let Value::Object(ref mut map) = payload {
        map.remove("signature");
    } else {
        return VerificationResult::invalid("AUTH_MALFORMED", "authorization payload must be an object");
    }

    let signing_input = match authorization_signing_input(&payload) {
        Ok(v) => v,
        Err(_) => {
            return VerificationResult::invalid("AUTH_MALFORMED", "failed to construct canonical signing input")
        }
    };

    let sig_bytes = match STANDARD.decode(&auth.signature) {
        Ok(v) => v,
        Err(_) => return VerificationResult::invalid("AUTH_SIGNATURE_INVALID", "signature verification failed"),
    };

    let signature = match Signature::from_slice(&sig_bytes) {
        Ok(v) => v,
        Err(_) => return VerificationResult::invalid("AUTH_SIGNATURE_INVALID", "signature verification failed"),
    };

    if key.verify(&signing_input, &signature).is_err() {
        return VerificationResult::invalid("AUTH_SIGNATURE_INVALID", "signature verification failed");
    }

    VerificationResult::ok()
}
